packages/squiggle-lang/src/public/SqProject/SqModule.ts

```diff
@@ -0,0 +1,180 @@
+import { parse } from "../../ast/parse.js";
+import { AST, LocationRange } from "../../ast/types.js";
+import { Env } from "../../dists/env.js";
+import { errMap, result } from "../../utility/result.js";
+import {
+  SqCompileError,
+  SqError,
+  SqImportError,
+  SqOtherError,
+} from "../SqError.js";
+import { SqLinker } from "../SqLinker.js";
+import { ModulePointer, ProjectState } from "./ProjectState.js";
+import { SqModuleOutput } from "./SqModuleOutput.js";
+import { getHash } from "./utils.js";
+
+export type Import = {
+  name: string;
+  hash: string | undefined;
+  variable: string;
+  location: LocationRange;
+};
+
+export function importToPointer(importBinding: Import): ModulePointer {
+  return {
+    name: importBinding.name,
+    hash: importBinding.hash,
+  };
+}
+
+type ImportModules =
+  | {
+      type: "loaded";
+      value: Record<string, SqModule>;
+    }
+  | {
+      type: "loading";
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+type ImportOutputs =
+  | {
+      type: "loaded";
+      value: Record<string, SqModuleOutput>;
+    }
+  | {
+      type: "loading"; // or running - i.e. ImportModules might be loaded, but outputs don't exist yet
+    }
+  | {
+      type: "failed";
+      value: SqError;
+    };
+
+export class SqModule {
+  name: string;
+  code: string;
+  // key is module name, value is hash
+  pins: Record<string, string>;
```

Comment:
Pins are technically supported, and `ProjectState` designed in a pins-friendly way (which is not trivial! see `modules` vs `resolutions` distinction, etc.), but tbh I have no idea if they work.

The current code in this PR doesn't use them, and adding them is extra work on Squiggle Hub side.

I could write tests, but it would take more time, and it's wasteful to write tests for the unused feature.

Reply:
Another obvious point to consider writing this sort of thing in the code directly